crates/bevy_ecs/src/entity/mod.rs

```diff
@@ -681,486 +676,328 @@ impl SparseSetIndex for Entity {
+    ) -> EntityIdLocation {
+        self.ensure_row(row);
+        // SAFETY: We just did `ensure_row`
+        self.update(row, location)
     }
 
-    /// Increments the `generation` of a freed [`Entity`]. The next entity ID allocated with this
-    /// `index` will count `generation` starting from the prior `generation` + the specified
-    /// value + 1.
-    ///
-    /// Does nothing if no entity with this `index` has been allocated yet.
-    pub(crate) fn reserve_generations(&mut self, index: u32, generations: u32) -> bool {
-        if (index as usize) >= self.meta.len() {
-            return false;
+    /// Ensures row is valid.
+    #[inline]
+    fn ensure_row(&mut self, row: EntityRow) {
+        #[cold] // to help with branch prediction
+        fn expand(meta: &mut Vec<EntityMeta>, len: usize) {
+            meta.resize(len, EntityMeta::FRESH);
+            // Set these up too while we're here.
+            meta.resize(meta.capacity(), EntityMeta::FRESH);
         }
 
-        let meta = &mut self.meta[index as usize];
-        if meta.location.is_none() {
-            meta.generation = meta.generation.after_versions(generations);
-            true
-        } else {
-            false
+        let index = row.index() as usize;
+        if self.meta.len() <= index {
+            // TODO: hint unlikely once stable.
+            expand(&mut self.meta, index + 1);
         }
     }
 
-    /// Get the [`Entity`] with a given id, if it exists in this [`Entities`] collection
-    /// Returns `None` if this [`Entity`] is outside of the range of currently reserved Entities
+    /// Marks the `row` as free, returning the [`Entity`] to reuse that [`EntityRow`].
     ///
-    /// Note: This method may return [`Entities`](Entity) which are currently free
-    /// Note that [`contains`](Entities::contains) will correctly return false for freed
-    /// entities, since it checks the generation
-    pub fn resolve_from_id(&self, row: EntityRow) -> Option<Entity> {
-        let idu = row.index() as usize;
-        if let Some(&EntityMeta { generation, .. }) = self.meta.get(idu) {
-            Some(Entity::from_raw_and_generation(row, generation))
-        } else {
-            // `id` is outside of the meta list - check whether it is reserved but not yet flushed.
-            let free_cursor = self.free_cursor.load(Ordering::Relaxed);
-            // If this entity was manually created, then free_cursor might be positive
-            // Returning None handles that case correctly
-            let num_pending = usize::try_from(-free_cursor).ok()?;
-            (idu < self.meta.len() + num_pending).then_some(Entity::from_raw(row))
+    /// # Safety
+    ///
+    /// - `row` must be destructed (have no location) already.
+    pub(crate) unsafe fn mark_free(&mut self, row: EntityRow, generations: u32) -> Entity {
+        // We need to do this in case an entity is being freed that was never constructed.
+        self.ensure_row(row);
```

Comment:
The name of this method is very unclear in context. I would prefer `ensure_row_is_SOMETHING` here.